### DISCUSSION TITLE
Fcitx5 works!

### DISCUSSION THREAD

**Discussion by JustSimplyKyle** [hyprwm/Hyprland#421] [Show and tell]
it's *barely working. 
`windowrule=pseudo,fcitx`
```
exec-once=fcitx5-remote -r
exec-once=fcitx5 -d --replace
exec-once=fcitx5-remote -r
```
`bind=ALT,E,exec,pkill fcitx5 -9;sleep 1;fcitx5 -d --replace; sleep 1;fcitx5-remote -r`
the last line is for resucing, since this method can be... how do I elabrate, unstable. You can sometimes make the input method screen shake??? and `wezterm` suddenly doesn't allow me to input stuff.
It seems like programs running under xwayland have a hard time working with fcitx5, since they have sorta a ring around it.
![image](https://user-images.githubusercontent.com/68589851/180637747-2773c5b8-06df-42ac-952e-3689ac8b18e1.png)
![image](https://user-images.githubusercontent.com/68589851/180637768-d0b21706-8ad1-474f-bc8b-a4ab39cbd9d2.png)

And often the times that it *shakes*, it's in an xwayland screen

**Comment by JustSimplyKyle** [comment]
The said video recording it *shaking*
https://user-images.githubusercontent.com/68589851/180637917-e1ae2c13-473e-4da1-be93-d01748517a81.mp4

**Reply by JustSimplyKyle** [reply]
and no, I can't just backspace it out, it's stuck in place and very annoying.
`windowrule=pseudo,fcitx` this *somewhat* makes it more stable, setting it to `floating` seems to make it more unstable

**Reply by JustSimplyKyle** [reply]
is it possible to just tell hyprland... don't move it?

**Reply by JustSimplyKyle** [reply]
ok.. it seems** like this happens when the input method box touch the edge and wanted to move, but hyprland says no!! so it got stucked between can it move, or can it not

**Reply by vaxerski** [reply]
in pseudo mode hyprland will reject any configure requests. In floating, the window will be moved wherever it wants to.

**Comment by realth000** [comment]
The latest hyprland with these packages works in terminal, chrome/firefox, wechat-uos, qq and vs code:

> fcitx5 5.0.23-1
fcitx5-breeze 2.0.0-1
fcitx5-chinese-addons 5.0.17-1
fcitx5-configtool 5.0.17-1
fcitx5-gtk 5.0.23-1
fcitx5-pinyin-sougou 20210320-1
fcitx5-pinyin-zhwiki 1:0.2.4.20230329-1
fcitx5-qt 5.0.17-1

But still can not use in Jetbrain IDE.

输入法都切不动，废了。

**Reply by EndCredits** [reply]
Have you tried its JetBrains IDE section.

大佬，这个关于 JB IDE 的试过了吗

https://wiki.archlinux.org/title/Fcitx5#Troubleshooting

**Reply by realth000** [reply]
> Have you tried its JetBrains IDE section.
> 
> 大佬，这个关于 JB IDE 的试过了吗
> 
> https://wiki.archlinux.org/title/Fcitx5#Troubleshooting

试了，没用。我这里的问题是切不了输入法，不是候选框位置不对。

**Reply by HyperCherry** [reply]
have u try set fcitx environment variables?
```
GTK_IM_MODULE=fcitx
QT_IM_MODULE=fcitx
XMODIFIERS=@im=fcitx
```
[Fcitx5-ArchWiki](https://wiki.archlinux.org/title/Fcitx5)

**Reply by realth000** [reply]
> have u try set fcitx environment variables?
> 
> ```
> GTK_IM_MODULE=fcitx
> QT_IM_MODULE=fcitx
> XMODIFIERS=@im=fcitx
> ```
> 
> [Fcitx5-ArchWiki](https://wiki.archlinux.org/title/Fcitx5)

yes, i already saved following in `~/.pam_environment`.
``` shell
✿ cat ~/.pam_environment
#GTK_IM_MODULE DEFAULT=fcitx
#QT_IM_MODULE  DEFAULT=fcitx
#XMODIFIERS    DEFAULT=\@im=fcitx
#SDL_IM_MODULE DEFAULT=fcitx
# Wayland compatibility
QT_QPA_PLATFORM         DEFAULT=wayland
CLUTTER_BACKEND         DEFAULT=wayland
SDL_VIDEODRIVER         DEFAULT=wayland
MOZ_ENABLE_WAYLAND      DEFAULT=1
MOZ_WEBRENDER           DEFAULT=1
XDG_SESSION_TYPE        DEFAULT=wayland
XDG_CURRENT_DESKTOP     DEFAULT=sway

# QT-related theming
QT_QPA_PLATFORMTHEME    DEFAULT=qt5ct

# FCITX input-related
#GLFW_IM_MODULE         DEFAULT=ibus
GLFW_IM_MODULE          DEFAULT=fcitx
GTK_IM_MODULE           DEFAULT=fcitx
INPUT_METHOD            DEFAULT=fcitx
XMODIFIERS              DEFAULT=@im=fcitx
IMSETTINGS_MODULE       DEFAULT=fcitx
QT_IM_MODULE            DEFAULT=fcitx
```

**Reply by HyperCherry** [reply]
`~/.pam_environment` seems has been deprecated and it's not read anymore.How about try other methods to set?
> [Using pam_env](https://wiki.archlinux.org/title/environment_variables#Using_pam_env)

**Reply by realth000** [reply]
> `~/.pam_environment` seems has been deprecated and it's not read anymore.How about try other methods to set?
> 
> > [Using pam_env](https://wiki.archlinux.org/title/environment_variables#Using_pam_env)

Now I set them in `/etc/environment` and `/etc/security/pam_en.conf` following the guide but nothing changes in JetBrains.

Jetbrains IDE runs on x11 (xwayland?) on my machine so this may be the cause? I also use neovide on x11 which also can not use fcitx5.

I'll check this later.

neovide failed to start if wayland enabled, not related to fcitx.

**Comment by realth000** [comment]
Thanks @HyperCherry finally fcitx5 works with JetBrain IDE on my machine.

Make sure these variables only exist in `/etc/security/pam_env.conf`.
Otherwise may be confused by part of them in `~/.zshrc`, `~/.pam_environment` or `/etc/profile`.

<details><summary>Click here to see configs in `/etc/security/pam_env.conf`</summary>

``` shell
#
# This is the configuration file for pam_env, a PAM module to load in
# a configurable list of environment variables for a
#
# The original idea for this came from Andrew G. Morgan ...
#<quote>
#   Mmm. Perhaps you might like to write a pam_env module that reads a
#   default environment from a file? I can see that as REALLY
#   useful... Note it would be an "auth" module that returns PAM_IGNORE
#   for the auth part and sets the environment returning PAM_SUCCESS in
#   the setcred function...
#</quote>
#
# What I wanted was the REMOTEHOST variable set, purely for selfish
# reasons, and AGM didn't want it added to the SimpleApps login
# program (which is where I added the patch). So, my first concern is
# that variable, from there there are numerous others that might/would
# be useful to be set: NNTPSERVER, LESS, PATH, PAGER, MANPAGER .....
#
# Of course, these are a different kind of variable than REMOTEHOST in
# that they are things that are likely to be configured by
# administrators rather than set by logging in, how to treat them both
# in the same config file?
#
# Here is my idea:
#
# Each line starts with the variable name, there are then two possible
# options for each variable DEFAULT and OVERRIDE.
# DEFAULT allows an administrator to set the value of the
# variable  to some default value, if none is supplied then the empty
# string is assumed. The OVERRIDE option tells pam_env that it should
# enter in its value (overriding the default value) if there is one
# to use. OVERRIDE is not used, "" is assumed and no override will be
# done.
#
# VARIABLE   [DEFAULT=[value]]  [OVERRIDE=[value]]
#
# (Possibly non-existent) environment variables may be used in values
# using the ${string} syntax and (possibly non-existent) PAM_ITEMs may
# be used in values using the @{string} syntax. Both the $ and @
# characters can be backslash escaped to be used as literal values
# values can be delimited with "", escaped " not supported.
# Note that many environment variables that you would like to use
# may not be set by the time the module is called.
# For example, HOME is used below several times, but
# many PAM applications don't make it available by the time you need it.
#
#
# First, some special variables
#
# Set the REMOTEHOST variable for any hosts that are remote, default
# to "localhost" rather than not being set at all
#REMOTEHOST	DEFAULT=localhost OVERRIDE=@{PAM_RHOST}
#
# Set the DISPLAY variable if it seems reasonable
#DISPLAY		DEFAULT=${REMOTEHOST}:0.0 OVERRIDE=${DISPLAY}
#
#
#  Now some simple variables
#
#PAGER		DEFAULT=less
#MANPAGER	DEFAULT=less
#LESS		DEFAULT="M q e h15 z23 b80"
#NNTPSERVER	DEFAULT=localhost
#PATH		DEFAULT=${HOME}/bin:/usr/local/bin:/bin\
#:/usr/bin:/usr/local/bin/X11:/usr/bin/X11
#
# silly examples of escaped variables, just to show how they work.
#
#DOLLAR		DEFAULT=\$
#DOLLARDOLLAR	DEFAULT=	OVERRIDE=\$${DOLLAR}
#DOLLARPLUS	DEFAULT=\${REMOTEHOST}${REMOTEHOST}
#ATSIGN		DEFAULT=""	OVERRIDE=\@
#
# This file is parsed by pam_env module
#
# Syntax: simple "KEY=VAL" pairs on separate lines
#
#
# XMODIFIRES=@im=fcitx
# QT_IM_MODULE=fcitx
# GTK_IM_MODULE=fcitx
# SDL_IM_MODULE=fcitx

#GTK_IM_MODULE DEFAULT=fcitx
#QT_IM_MODULE  DEFAULT=fcitx
#XMODIFIERS    DEFAULT=\@im=fcitx
#SDL_IM_MODULE DEFAULT=fcitx
# Wayland compatibility
QT_QPA_PLATFORM         DEFAULT=wayland
CLUTTER_BACKEND         DEFAULT=wayland
SDL_VIDEODRIVER         DEFAULT=wayland
MOZ_ENABLE_WAYLAND      DEFAULT=1
MOZ_WEBRENDER           DEFAULT=1
XDG_SESSION_TYPE        DEFAULT=wayland
XDG_CURRENT_DESKTOP     DEFAULT=sway

# QT-related theming
QT_QPA_PLATFORMTHEME    DEFAULT=qt5ct

# FCITX input-related
#GLFW_IM_MODULE         DEFAULT=ibus
GLFW_IM_MODULE          DEFAULT=fcitx
GTK_IM_MODULE           DEFAULT=fcitx
INPUT_METHOD            DEFAULT=fcitx
XMODIFIERS              DEFAULT=@im=fcitx
IMSETTINGS_MODULE       DEFAULT=fcitx
QT_IM_MODULE            DEFAULT=fcitx
SDL_IM_MODULE           DEFAULT=fcitx

```

</details>

Then reboot. Exit hyprland and tty seems not enough.

**Reply by muttnooob** [reply]
Is this still working for anybody?  Would you care sharing the key steps to enabling this on Hyprland? (I'm on NixOS, hopefully it works on NixOS Hyprland for someone)  Much appreciated!

**Reply by janik-git** [reply]
Using 
```nix
  i18n.inputMethod = {
    enable = true;
    type = "fcitx5";
    fcitx5.addons = with pkgs; [
      fcitx5-hangul
      # fcitx5-gtk
    ];
  };
```
in my nix config, 
and 
```
exec-once=fcitx5-remote -r
exec-once=fcitx5 -d --replace &
exec-once=fcitx5-remote -r
bind=SUPER_CTRL,space,execr,fcitx5-remote -t
```
works for me on Hyprland 0.45.2, and Nixos 24.11

**Reply by MattBrooks95** [reply]
I had a hard time getting it working this week on Nixos 24.11 for Japanese input with mozc.

I'm not sure which step fixed it but this link https://github.com/fcitx/fcitx5/discussions/1208 pointed out that I don't (and probably shouldn't) install fcitx5 in systemPackages.

Then, this bit in this very thread `windowrule=pseudo,fcitx` was missing from my hyprland config. I'm not sure how it was working before, but referencing these links and following the advice got it to work for me.

I'm still not good at Nix but this is my input_method_settings.nix module
```nix
{ config, pkgs, ... }:{
  i18n = {
    inputMethod = {
      type = "fcitx5";
      enable = true;
      fcitx5 = {
        waylandFrontend = true;
        addons = with pkgs; [
          fcitx5-mozc
          fcitx5-gtk
        ];
      };
    };
  };
}
```

And the section of my hypr_config.nix:
```nix
${if startFcitx5 then
''
#### interpolated from nix ###
# IME for Japanese. Will need mozc and fcitx5 to exist
windowrule = pseudo, fcitx
exec-once = fcitx5 -d -r
exec-once = fcitx5-remote -r
##############################
'' else ""}
```

**Comment by buildding** [comment]
You can try to add parameters when starting vscode and typora:
--enable-features=UseOzonePlatform --ozone-platform=wayland --enable-wayland-ime
or
Create a new file in  ~/.config/typora-flags.conf，or ~/.config/code-flags.conf 
Write the above parameters and save them，You can use fcitx5 in code or typora！！

**Comment by szchan** [comment]
Old electron apps have this problem.

You can try adding: "--enable-features=UseOzonePlatform --ozone-platform=wayland --enable-wayland-ime" when you start these apps.

Just like use "obsidian --enable-features=UseOzonePlatform --ozone-platform=wayland --enable-wayland-ime" to start obsidian.

If it worked for you, you can add "

--enable-features=UseOzonePlatform
--ozone-platform=wayland
--enable-wayland-ime

" into "~/.config/electron-flags.conf".

Then your ime will work as it should do in the old electron apps. Just close and start your electron apps again.
